cli/tests/test_next_prev_commands.rs

```diff
@@ -60,40 +62,86 @@ fn test_next_multiple() {
     let repo_path = test_env.env_root().join("repo");
     test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "first"]);
     test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "second"]);
     test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "third"]);
+    // @ is an empty discardable commit created by `jj commit`. It will be
+    // abandoned after we move with `jj prev`.
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @  zsuskulnrvyr
+    ◉  kkmpptxzrspx third
+    ◉  rlvkpnrzqnoo second
+    ◉  qpvuntsmwlqt first
+    ◉  zzzzzzzzzzzz
+    "###);
+
     let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["prev"]);
-    // The working copy commit is now a child of "second".
+    // The working copy commit is now a child of "second", even though the
+    // commit before our empty working copy commit is actually "third". This is
+    // because we skip the empty discardable @ commit when deciding how far to
+    // move backwards.
     insta::assert_snapshot!(stdout, @"");
     insta::assert_snapshot!(stderr, @r###"
-    Working copy now at: mzvwutvl f7a0f876 (empty) (no description set)
+    Working copy now at: royxmykx 5647d685 (empty) (no description set)
     Parent commit      : rlvkpnrz 5c52832c (empty) second
     "###);
 }
 
+#[test]
+fn test_prev_editing_non_discardable_leaf_revision() {
+    // Move @- from second to first. This means the new @ revision will be a
+    // child of the first commit.
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
+    test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "first"]);
+    test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "second"]);
+    test_env.jj_cmd_ok(&repo_path, &["describe", "-m", "third"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @  kkmpptxzrspx third
+    ◉  rlvkpnrzqnoo second
+    ◉  qpvuntsmwlqt first
+    ◉  zzzzzzzzzzzz
+    "###);
+
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["prev"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Working copy now at: royxmykx f04d8595 (empty) (no description set)
+    Parent commit      : qpvuntsm 69542c19 (empty) first
+    "###);
+}
+
 #[test]
 fn test_prev_multiple_without_root() {
-    // Move from fourth => second.
+    // Move @- from third to first.
     let test_env = TestEnvironment::default();
     test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
     let repo_path = test_env.env_root().join("repo");
     test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "first"]);
     test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "second"]);
     test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "third"]);
     test_env.jj_cmd_ok(&repo_path, &["commit", "-m", "fourth"]);
+    test_env.jj_cmd_ok(&repo_path, &["edit", "description(fourth)"]);
```

Comment:
The comment on the top originally said
```
// Move from fourth => second.
```

Which I took to mean "move `@` from fourth to second", so I added the edit.

I suppose it actually means move `@-` from fourth to second. I'll change it back so that `@` is an empty leaf.